src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs

```diff
@@ -56,8 +58,351 @@ public class BotSegmentsPlaybackController : MonoBehaviour
+                        // but if you go too far.. you can match segments in the replay that you won't see for another 50 segments when you go back to the menu again.. which is obviously wrong
+                        var lastSegment = _nextBotSegments[^1];
+                        if (lastSegment.Replay_TransientMatched)
+                        {
+                            if (_nextBotSegments.Count < 2)
+                            {
+                                var next = _dataPlaybackContainer.DequeueBotSegment();
+                                if (next != null)
+                                {
+                                    _lastTimeLoggedKeyFrameConditions = now;
+                                    FindObjectOfType<ReplayToolbarManager>()?.SetKeyFrameWarningText(null);
+                                    RGDebug.LogInfo($"({next.Replay_SegmentNumber}) - Bot Segment - Added {(next.HasTransientCriteria ? "" : "Non-")}Transient BotSegment for Evaluation after Transient BotSegment - {next.name ?? next.resourcePath} - {next.description}");
+                                    _nextBotSegments.Add(next);
+                                    //next while loop iteration will get this guy
+                                }
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // segment list empty.. dequeue another
+                        var next = _dataPlaybackContainer.DequeueBotSegment();
+                        if (next != null)
+                        {
+                            _lastTimeLoggedKeyFrameConditions = now;
+                            FindObjectOfType<ReplayToolbarManager>()?.SetKeyFrameWarningText(null);
+                            RGDebug.LogInfo($"({next.Replay_SegmentNumber}) - Bot Segment - Added {(next.HasTransientCriteria ? "" : "Non-")}Transient BotSegment for Evaluation - {next.name ?? next.resourcePath} - {next.description}");
+                            _nextBotSegments.Add(next);
+                            //next while loop iteration will get this guy
+                        }
+                    }
+                }
+
+                if (matchedThisUpdate)
+                {
+                    // only do this when a segment passed this update after all segments have been considered for this update
+                    KeyFrameEvaluator.Evaluator.PersistPriorFrameStatus();
+                }
+            }
+            catch (Exception ex)
+            {
+                var loggedMessage = $"(?) - Bot Segment - Exception processing BotSegments\r\n" + ex.Message;
+                LogPlaybackWarning(loggedMessage, ex);
+                // uncaught exception... stop the segment
+                UnloadSegmentsAndReset();
+                throw;
+            }
+        }
+
+        /**
```

Comment:
This is where the exploration magic happens on playback / the algorithm from Update -> Update is documented